### PR TITLE
bears/python: Add PyFlakesBear

### PR DESCRIPTION
**Pull request by srisankethu**
Closes https://github.com/coala/coala-bears/issues/825

### CODE REVIEW

File: requirements.txt [hunk 1]
```diff
@@ -28,6 +28,6 @@ pyyaml==3.*
 vulture==0.10.*
 bandit==1.1.*
 nbformat>=4.*
-pyflakes==1.2.*  # Although we don't need this directly, solves a dep conflict
+pyflakes==1.2.*
```
Comment (srisankethu) [inline requirements.txt]: Should I change it to version 1.3.* ?

File: bears/python/PyFlakesBear.py [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+    Checks Python files for logistic errors using ``pyflakes``.
+
+    See https://github.com/PyCQA/pyflakes for more info.
+    """
+    LANGUAGES = {"Python", "Python 3"}
+    REQUIREMENTS = {PipRequirement('pyflakes', '1.12.*')}
+    AUTHORS = {'The coala developers'}
+    AUTHORS_EMAILS = {'coala-devel@googlegroups.com'}
+    LICENSE = 'AGPL-3.0'
+    ASCIINEMA_URL = 'https://asciinema.org/a/'
```
Comment (srisankethu) [inline bears/python/PyFlakesBear.py]: I will make an asciinema once the bear is ready.

File: bears/python/PyFlakesBear.py [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+        output_format='regex',
+        output_regex=r'(?P<filename>.*):(?P<line>.*):\s*(?P<message>.*)')
+class PyFlakesBear:
+    """
+    Checks Python files for logistic errors using ``pyflakes``.
+
+    See https://github.com/PyCQA/pyflakes for more info.
+    """
+    LANGUAGES = {"Python", "Python 3"}
+    REQUIREMENTS = {PipRequirement('pyflakes', '1.12.*')}
```
Comment (srisankethu) [inline bears/python/PyFlakesBear.py]: Change version to 1.3.* ?

File: requirements.txt (hunk 1 above)
Comment (jayvdb) [inline requirements.txt]: Ya. Bump up. Lovely dictionary dup checks in 1.3.

**Comment by jayvdb** [comment]
We need a severity map, at least defining [`UnusedImport`](https://github.com/PyCQA/pyflakes/blob/master/pyflakes/messages.py) as only info, not error.
That is the most common "false positive", often found when `__all__` is not defined, as it is not necessary.
`ImportStarUsed` and `ImportStarUsage` are also not errors.
`UnusedVariable` is also not an error, but could be a clue of a bug; I would leave this as an error.

**Comment by srisankethu** [comment]
@jayvdb By default RESULTY_SEVERITY.NORMAL is used which is a "warn".
Without using a severity_map as there are not error codes, there are no other options in ``@linter`` to use.

**Comment by jayvdb** [comment]
Yea, that is why I recommend using the pyflakes API, not the command line tool.  See e.g. the flake8 code that wraps pyflakes. It will be Python 3 only, but that is how flake8 works also. And saying a bear that uses Python ast supports Python 2 and 3 is not possible (https://github.com/coala/coala-bears/issues/929).  It can only support one at a time, as the program name is an executable script that is either Python 2 or Python 3, and coala has no idea which version it is.

**Comment by sils** [comment]
@srisankethu removing pending review label and adding blocked. Be sure to ping us, nobody will review this if the pending revew isn't readded ;)

**Comment by srisankethu** [comment]
@jayvdb Like you have suggested on the gitter channel, all the errors will be taken as INFO

**Comment by sils** [comment]
@srisankethu tests fail?

**Comment by srisankethu** [comment]
@sils The tests are failing because of the pyflakes version - ``1.3.*``. It seems that ``1.2.*`` is required for flake8 -> https://circleci.com/gh/coala/coala-bears/3774?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link.

File: requirements.txt (hunk 1 above)
Comment (sils) [inline requirements.txt]: add a comment why this isn't 1.3 please if it can't be 1.3

File: bears/python/PyFlakesBear.py [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+@linter(executable='pyflakes',
+        output_format='regex',
+        output_regex=r'(?P<filename>.*):(?P<line>.*):\s*'
+                     r'(?P<severity>)\s(?P<message>.*)',
+        severity_map={
+            '': RESULT_SEVERITY.INFO
+        })
+class PyFlakesBear:
+    """
+    Checks Python files for logistic errors using ``pyflakes``.
```
Comment (sils) [inline bears/python/PyFlakesBear.py]: logistic? tat's not the right word for sure

File: bears/python/PyFlakesBear.py [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+    LANGUAGES = {"Python", "Python 3"}
+    REQUIREMENTS = {PipRequirement('pyflakes', '1.2.*')}
+    AUTHORS = {'The coala developers'}
+    AUTHORS_EMAILS = {'coala-devel@googlegroups.com'}
+    LICENSE = 'AGPL-3.0'
+    ASCIINEMA_URL = 'https://asciinema.org/a/92503'
+    CAN_DETECT = {'Unused imports', 'Undefined Element'}
+
+    @staticmethod
+    def create_arguments(filename, file, config_file):
```
Comment (sils) [inline bears/python/PyFlakesBear.py]: what about turning on and off checks here?

**Comment by sils** [comment]
unack https://github.com/coala/coala-bears/pull/985/commits/a02cfa2229953eecc5595b56be2a1d00a1efb2d5

File: requirements.txt (hunk 1 above)
Comment (srisankethu) [inline requirements.txt]: The tests failed because of flake8 requirements when 1.3.* is used.

File: bears/python/PyFlakesBear.py (hunk 4 above)
Comment (srisankethu) [inline bears/python/PyFlakesBear.py]: http://blog.sideci.com/2015/12/14/style-guide-of-python-and-lint-tool/ says logistic errors. Should I simply make it errors then?

File: bears/python/PyFlakesBear.py [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+
+    See https://github.com/PyCQA/pyflakes for more info.
+    """
+    LANGUAGES = {"Python", "Python 3"}
+    REQUIREMENTS = {PipRequirement('pyflakes', '1.2.*')}
+    AUTHORS = {'The coala developers'}
+    AUTHORS_EMAILS = {'coala-devel@googlegroups.com'}
+    LICENSE = 'AGPL-3.0'
+    ASCIINEMA_URL = 'https://asciinema.org/a/92503'
+    CAN_DETECT = {'Unused imports', 'Undefined Element'}
```
Comment (srisankethu) [inline bears/python/PyFlakesBear.py]: Yes!

File: requirements.txt (hunk 1 above)
Comment (Nosferatul) [inline requirements.txt]: Try with 1.3.*. See if it pass the tests

File: bears/python/PyFlakesBear.py (hunk 4 above)
Comment (Nosferatul) [inline bears/python/PyFlakesBear.py]: just `errors` should be enough

File: bears/python/PyFlakesBear.py [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+from coalib.bears.requirements.PipRequirement import PipRequirement
+from coalib.results.RESULT_SEVERITY import RESULT_SEVERITY
+
+
+@linter(executable='pyflakes',
+        output_format='regex',
+        output_regex=r'(?P<filename>.*):(?P<line>.*):\s*'
+                     r'(?P<severity>)\s(?P<message>.*)',
+        severity_map={
+            '': RESULT_SEVERITY.INFO
```
Comment (AbdealiJK) [inline bears/python/PyFlakesBear.py]: ?? Nothing in the severity map ?

**Comment by AbdealiJK** [comment]
Looks good - just the 1.3.x issue and the severity_map is an issue as far as I see (y)

**Comment by AbdealiJK** [comment]
Oh. just read @jayvdb 's comment on the Python API. @jayvdb Maybe we can get this merged and create an issue for the Python API ? (Something's better than nothing and pyflakes is powerful)

**Comment by jayvdb** [comment]
Yes. If all errors are 'info' now, we can merge this now. Then look at ways to set only some messages to error.

File: requirements.txt (hunk 1 above)
Comment (srisankethu) [inline requirements.txt]: The requirement for flake8 seems to be a version < 1.3.0 -> [test failure](https://circleci.com/gh/coala/coala-bears/3774?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link)

**Comment by srisankethu** [comment]
@jayvdb Should I make it syntax detectable too. As syntax errors are not INFO.

**Comment by jayvdb** [comment]
SyntaxError can be level error. That wont change in the future.

**Comment by srisankethu** [comment]
@jayvdb So should I make it detectable? https://github.com/coala/coala-bears/pull/985#issuecomment-260171486

File: bears/python/PyFlakesBear.py (hunk 4 above)
Comment (jayvdb) [inline bears/python/PyFlakesBear.py]: "erroneous and ambiguous syntax" covers pyflakes pretty accurately. Ive not seen a clearer definition. The maintainers have a goal of only reporting errors, but that doesnt explain 50% of the messages.

File: requirements.txt (hunk 1 above)
Comment (jayvdb) [inline requirements.txt]: Where/why is the flake8 package being added? We dont have a flake8 bear yet...?

**Comment by jayvdb** [comment]
Ya, it will be perfect if you can make SyntaxError an error, and everything else is info.

**Comment by srisankethu** [comment]
@jayvdb I have tried doing that by changing my regex, but it seems that when it introduce ``(?P<column>)`` it is assigned ``''`` which is raising an error since it is not a int. For now this bear will detect ``Unused imports`` and ``Undefined elements``.

**Comment by jayvdb** [comment]
Then do info level for everything, and polish the patch to get it merged.

**Comment by srisankethu** [comment]
@jayvdb Every error is info now. ``Syntax error`` is not detected because it has a different ouput.

File: requirements.txt (hunk 1 above)
Comment (jayvdb) [inline requirements.txt]: Maybe circle preinstalls flake8?if so, uninstall it

**Comment by jayvdb** [comment]
How is it different?

**Comment by srisankethu** [comment]
https://asciinema.org/a/92541

**Comment by jayvdb** [comment]
Ah. Discard the lines which are from the source. Coala does need that.

File: bears/python/PyFlakesBear.py (hunk 5 above)
Comment (sils) [inline bears/python/PyFlakesBear.py]: config file?

Comment (jayvdb) [inline bears/python/PyFlakesBear.py]: no config.  IIRC, only one env var, to enable checking of doctest .

File: bears/python/PyFlakesBear.py (hunk 6 above)
Comment (jayvdb) [inline bears/python/PyFlakesBear.py]: This should be "Unused Code", same as https://github.com/coala/coala-bears/blob/master/bears/python/PyUnusedCodeBear.py#L15 ?

Comment (srisankethu) [inline bears/python/PyFlakesBear.py]: @sils Well its with the regex. pyflakes gives a different output while checking Syntax errors -> https://asciinema.org/a/92541. May be I am bad with regex. I will update the PR.

File: bears/python/PyFlakesBear.py [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+
+
+@linter(executable='pyflakes',
+        use_stderr=True,
+        output_format='regex',
+        output_regex=r'(?P<filename>.*):(?P<line>\d+):'
+                     r'(?P<column>\d*):*(?P<severity>)\s(?P<message>.*)\n',
+        severity_map={
+            '': RESULT_SEVERITY.INFO
+        })
```
Comment (sils) [inline bears/python/PyFlakesBear.py]: if there's no severity, just remove the group from the regex

**Comment by sils** [comment]
Otherwise it seems good

**Comment by sils** [comment]
though you should get the tests to pass

File: bears/python/PyFlakesBear.py (hunk 8 above)
Comment (srisankethu) [inline bears/python/PyFlakesBear.py]: @jayvdb Wants all the errors to be addressed as INFO. But by default all the errors are logged to NORMAL.

**Comment by srisankethu** [comment]
@sils The regex looks fine right? But the thing is, its not working :( The output of PyFlakes is different in case of 'Syntax' checking. Specifying (?P<column>) expects an `int`. But it is receiving `''` - a `str` when errors other than 'Syntax' checking are checked. Also, it tried using `use_stderr=True` but it seems the bear isn't catching the ouput. This asciinema will make it clear -> https://asciinema.org/a/92541.
The tests will pass when the bear works properly!

File: requirements.txt (hunk 1 above)
Comment (jayvdb) [inline requirements.txt]: The pyflakes line was removed from requirements.txt, so you now need to re-add it, as `~=1.3` please.

**Comment by Mixih** [comment]
>'' - a str when errors other than 'Syntax' checking are checked

use the `|` regex syntax and check it before passing on?

**Comment by srisankethu** [comment]
@Mixih Thanks! I am bad with regex :stuck_out_tongue_winking_eye:

File: requirements.txt [hunk 9]
```diff
@@ -26,6 +26,7 @@ pyyaml~=3.12
 vulture~=0.10
 bandit~=1.2
 nbformat>=4.1
+pyflakes==1.2.*
```
Comment (nemaniarjun) [inline requirements.txt]: use the `~=` symbol

File: bears/python/PyFlakesBear.py [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+    LANGUAGES = {'Python', 'Python 3'}
+    REQUIREMENTS = {PipRequirement('pyflakes', '1.2.*')}
+    AUTHORS = {'The coala developers'}
+    AUTHORS_EMAILS = {'coala-devel@googlegroups.com'}
+    LICENSE = 'AGPL-3.0'
+    ASCIINEMA_URL = 'https://asciinema.org/a/92503'
+    CAN_DETECT = {'Syntax', 'Unused Code', 'Undefined Element'}
+
+    @staticmethod
+    def create_arguments(filename, file, config_file):
```
Comment (Mixih) [inline bears/python/PyFlakesBear.py]: PyFlakes doesn't accept a config_file, it only accepts a env variable! See @jayvdb's previous comment

**Comment by srisankethu** [comment]
Sorry! had conflict merge conflicts -> PR at https://github.com/coala/coala-bears/pull/1061

**Pull request closed without merging**